design-documents/promises.md

```diff
@@ -0,0 +1,240 @@
+### Why?
+Service contracts in the future will often be executed in an asynchronous manner
+and it's time to introduce a standard Promise to Magento for asynchronous operations to employ
+### Requirements
+* Promises CANNOT be forwarded with _then_ and _otherwise_ ([see explanation](#forwarding))
+* _then_ accepts a function that will be executed when the promise is resolved, the callback will
+receive a single argument - result of the execution
+* _otherwise_ accepts a function that will be executed if an error occurs during the asynchronous
+operation, it will receive a single argument - a _Throwable_
+* Promises can be used in a synchronous way to prevent methods that use methods returning promises
+having to return a promise as well ([see explanation](#callback-hell)); This will be done by promises having _wait_ method
+* _wait_ method does not throw an exception if the promise is rejected
+nor does it return the result of the resolved promise ([see explanation](#wait-not-unwrapping-promises))
+* If an exception occurs during an asynchronous operation and no _otherwise_ callback is
+provided then it will just be rethrown
+### API
+##### Promise to be used by client code
+When client code receives a promise from calling another object's method
+it shouldn't have access to _resolve_ and _reject_ methods, it should only be able to
+provide callbacks to process promised results and to wait for promised operation's execution.
+ 
+This interface will be used as the return type of methods returning promises.
+```php
+interface PromiseInterface
+{
+    /**
+     * @throws PromiseProcessedException When callback was alredy provided.
+     */
+    public function then(callable $callback): void;
+    
+    /**
+     * @throws PromiseProcessedException When callback was alredy provided.
+     */
+    public function otherwise(callable $callback): void;
+    
+    public function wait(): void;
+}
+```
+##### Promise to be created
+This promise will be created by asynchronous code
+```php
+interface ResultPromiseInterface extends PromiseInterface
+{
+    public function resolve($value): void;
+    
+    public function reject(\Throwable $exception): void;
+}
+```
+
+### Implementation
+A wrapper around [Guzzle Promises](https://github.com/guzzle/promises) will be created to implement the APIs above. Guzzle Promises fit
+most important criteria - they allow synchronous execution as well as asynchronous. It's a mature
```

Comment:
What are the benefits of the interface that would allow us to call something both ways?

Reply:
if a promise would only allow 'then' and 'otherwise' callback then all the client code using would have to be returning promises as well because there would be no way to actually wait for an async operation to finish